### PR TITLE
feature/renko-series-type

### LINKED ISSUE #12524: New Series Type: Renko
Implement Renko series type for Highstock.

### PR DESCRIPTION
**Pull request by pawellysy**
Added new series type, [Renko](https://www.highcharts.com/docs/stock/renko).
___
Created new series type: `Renko`. Closes #12524

### CODE REVIEW

File: ts/Series/Renko/RenkoSeriesDefaults.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+/* *
+ *
+ *  API Options
+ *
+ * */
+
+/**
+ * An Renko series is a style of financial chart used to describe price
+ * movements over time. It displays open, high, low and close values per
+ * data point.
```
Comment (pawelfus) [inline ts/Series/Renko/RenkoSeriesDefaults.ts:31]: I think Renko focuses only on close prices.

File: ts/Series/Renko/RenkoSeriesDefaults.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+ *    also be omitted, in which case the inner arrays should be of length 4.
+ *    Then the `x` value is automatically calculated, either starting at 0 and
+ *    incremented by 1, or from `pointStart` and `pointInterval` given in the
+ *    series options.
+ *    ```js
+ *    data: [
+ *        [0, 7, 2, 0, 4],
+ *        [1, 1, 4, 2, 8],
+ *        [2, 3, 3, 9, 3]
+ *    ]
```
Comment (pawelfus) [inline ts/Series/Renko/RenkoSeriesDefaults.ts:97]: Let's go with x+y values only.

File: ts/Series/Renko/RenkoSeriesOptions.d.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+     * is that the points are calculated based on the trends and boxSize.
+     * Setting options for indivcidual point is impossible.
+     *
+     *    ```js
+     *    data: [{
+     *        x: 1,
+     *        close: 6
+     *    }, {
+     *        x: 1,
+     *        close: 7,
```
Comment (pawelfus) [inline ts/Series/Renko/RenkoSeriesOptions.d.ts:88]: Let's rename to `y`, so we follow other cartesian series

File: ts/Series/Renko/RenkoSeries.ts [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        }
+    }
+
+    /* *
+     *
+     *  Static Properties
+     *
+     * */
+    public static compose(AxisClass: typeof Axis): void {
+        if (U.pushUnique(composed, 'pointandfigure')) {
```
Comment (pawelfus) [inline ts/Series/Renko/RenkoSeries.ts:148]: Renko*

**Review by pawelfus — CHANGES_REQUESTED** [review]

File: ts/Series/Renko/RenkoSeriesDefaults.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+import type RenkoSeriesOptions from './RenkoSeriesOptions';
+
+/* *
+ *
+ *  API Options
+ *
+ * */
+
+/**
+ * An Renko series is a style of financial chart used to describe price
```
Comment (DJTechnoo) [inline ts/Series/Renko/RenkoSeriesDefaults.ts:29]:
```suggestion
 * A Renko series is a style of financial chart used to describe price
```

File: ts/Series/Renko/RenkoSeriesOptions.d.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+import type ColorType from '../../Core/Color/ColorType';
+
+/* *
+ *
+ *  Declarations
+ *
+ * */
+
+/**
+ * An Renko series is a style of financial chart used to describe price
```
Comment (DJTechnoo) [inline ts/Series/Renko/RenkoSeriesOptions.d.ts:29]:
```suggestion
 * A Renko series is a style of financial chart used to describe price
```

**Review by DJTechnoo — CHANGES_REQUESTED** [review]
Looks very good! Found a few typos

File: docs/stock/renko.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+## Data Grouping and Tooltip Customization
+
+Highcharts optimizes viewer experience with large datasets by providing data grouping options. However, the dataGrouping is disabled for renko series, because it does not make sense to group those points by a unit of time.
+
+Moreover, you can provide detailed insights about each data point using customizable tooltips:
+
+```javascript
+tooltip: {
+    pointFormat: ' style="color:{point.color}">\u25CF</span> {series.name}: <b>{point.y}<br/>'
```
Comment (bm64) [inline docs/stock/renko.md:67]: Missing starting `span` tag

**Review by bm64 — CHANGES_REQUESTED** [review]

File: docs/stock/renko.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+The Heikin-Ashi technique differs from Candlestick and Renko methods by using the average prices to provide smoother trend lines, which might be more useful in catching longer-term trends. 
+
+Even though Renko charts simplify the representation, they can simplify trend identification as compared with the other two types.
+
+
+<iframe src="https://www.highcharts.com/samples/stock/demo/renko-vs-heikinashi-vs-candlestick/" width="100%" height="100%" title="Renko Chart"></iframe>
+
+## Data Grouping and Tooltip Customization
+
+Highcharts optimizes viewer experience with large datasets by providing data grouping options. However, the dataGrouping is disabled for renko series, because it does not make sense to group those points by a unit of time.
```
Comment (pawelfus) [inline docs/stock/renko.md:61]: dataGrouping -> `dataGrouping`

File: docs/stock/renko.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+```javascript
+tooltip: {
+    pointFormat: ' style="color:{point.color}">\u25CF</span> {series.name}: <b>{point.y}<br/>'
+}
+```
+## Summary
+
+Highchart's Renko series provides a distinct way of visualizing price data. It emphasized considerable price changes over a specified amount, disregarding time and minor fluctuations. This makes it a valuable tool for traders seeking to identify significant market trends, and it is but one of many useful charting techniques supported by Highcharts.
+
+STEP 3: Applying tone of voice rules.
```
Comment (pawelfus) [inline docs/stock/renko.md:74]: Nice

File: docs/stock/renko.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+FINAL_RESULT:# Highcharts: Introducing Renko Series
```
Comment (pawelfus) [inline docs/stock/renko.md:76]: Nice

File: docs/stock/renko.md [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+plotOptions: {
+    renko: {
+        boxSize: 1   // This indicates each "brick" symbolises $1 price alterations
+    }
+}
+```
+
+## Summary
+
+The Renko series that Highcharts provides is a valuable way of visualizing price data. By focusing on significant price changes and omitting time and minor swings, it becomes a handy resource for traders keen on identifying crucial market trends. It's just one of the many beneficial charting techniques supported by Highcharts.
```
Comment (pawelfus) [inline docs/stock/renko.md:107]: It seems we have two articles in one :)

File: samples/stock/demo/renko-vs-heikinashi-vs-candlestick/demo.details [hunk 12]
```diff
@@ -0,0 +1,11 @@
+---
+name: Heikin Ashi
+authors:
+  - Karol Kolodziej
```
Comment (pawelfus) [inline samples/stock/demo/renko-vs-heikinashi-vs-candlestick/demo.details:4]: Who?

File: samples/stock/demo/renko-vs-heikinashi-vs-candlestick/demo.details [hunk 13]
```diff
@@ -0,0 +1,11 @@
+---
+name: Heikin Ashi
```
Comment (pawelfus) [inline samples/stock/demo/renko-vs-heikinashi-vs-candlestick/demo.details:2]: Name

File: samples/stock/demo/renko-vs-heikinashi-vs-candlestick/demo.html
Comment (pawelfus) [inline samples/stock/demo/renko-vs-heikinashi-vs-candlestick/demo.html:1]: This should be in samples, not demos

File: samples/stock/demo/stock-tools-gui/demo.js
Comment (pawelfus) [inline samples/stock/demo/stock-tools-gui/demo.js:1]: Unexpected changes?

File: ts/Series/Renko/RenkoSeriesOptions.d.ts (hunk 3 above)
Comment (pawelfus) [reply ts/Series/Renko/RenkoSeriesOptions.d.ts:95]: Still incorrect.

File: ts/Series/Renko/RenkoSeriesOptions.d.ts [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+     *        x: 1,
+     *        close: 7,
+     *    }]
+     *    ```
+     *
+     * @type {Array<Array<(number|string),number,number,number>|Array<(number|string),number,number,number,number>|*>}
+     *
+     * @extends series.column.data
+     *
+     * @excluding y
```
Comment (pawelfus) [inline ts/Series/Renko/RenkoSeriesOptions.d.ts:103]: Why excluding `y`?

**Review by pawelfus — CHANGES_REQUESTED** [review]
In addition to the comments below:

1. It seems the API ref is missing: 
image

File: docs/stock/renko.md [hunk 15]
```diff
@@ -0,0 +1,107 @@
+# Highcharts: Understanding Renko Series
+
+Highcharts offers a variety of charting methods for visualizing price data, and one of these is the `Renko` series.
+
+Renko, a Japanese term that signifies "brick", is a charting method commonly used for technical securities trading. The unique aspect of [Renko charts](https://www.investopedia.com/terms/r/renkochart.asp) is that they do not factor in time. Instead, they solely focus on price changes that meet a predetermined amount, filling a "price box".
```
[the remark this answers is not in the document]
Comment (pawellysy) [reply docs/stock/renko.md:5]: we do not, but skald suggested that 😓

File: docs/stock/renko.md (hunk 10 above)
Comment (pawellysy) [reply docs/stock/renko.md:76]: 😓

File: samples/stock/demo/renko-vs-heikinashi-vs-candlestick/demo.html
Comment (pawellysy) [reply samples/stock/demo/renko-vs-heikinashi-vs-candlestick/demo.html:1]: any particular dir it should be in?

**Review by DJTechnoo — APPROVED** [review]
LGTM! Great job

**Review by bm64 — APPROVED** [review]
lgtm

**Review by pawelfus — CHANGES_REQUESTED** [review]
A few API requests:

- [ ] Options from API ref to remove:
- boost (and all derivatives)
- cumulative (and all derivatives)
- dataSorting
- dragDrop

- [x] Some of the options are removed from `plotOptions.renko` but still available under `series.renko` (eg http://localhost:9005/highstock/series.renko.dataGrouping).

- [x] A narrow chart is missing the navigator series - could it be an issue with DG? Demo: https://jsfiddle.net/Ln2gs69z/

- [x] Colors (up/down) are missing in styledMode: https://jsfiddle.net/Ln2gs69z/1/ 

- [x] `setData` seems to be broken: https://jsfiddle.net/Ln2gs69z/3/

File: samples/stock/demo/renko-vs-heikinashi-vs-candlestick/demo.details [hunk 16]
```diff
@@ -0,0 +1,11 @@
+---
+name: renko vs Heikin Ashi vs Candlestick
+authors:
+  - Pawel lysy
```
Comment (kamil-musialowski) [inline samples/stock/demo/renko-vs-heikinashi-vs-candlestick/demo.details:4]:
```suggestion
name: Renko vs Heikin Ashi vs Candlestick
authors:
  - Pawel Lysy
```

File: samples/stock/plotoptions/renko-boxsize/demo.details [hunk 17]
```diff
@@ -0,0 +1,6 @@
+---
+name: Renko series
+authors:
+  - Pawel lysy
```
Comment (kamil-musialowski) [inline samples/stock/plotoptions/renko-boxsize/demo.details:4]:
```suggestion
name: Renko series
authors:
  - Pawel Lysy
```

File: ts/Series/Renko/RenkoSeriesOptions.d.ts [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+import type ColorType from '../../Core/Color/ColorType';
+
+/* *
+ *
+ *  Declarations
+ *
+ * */
+
+/**
+ * An Renko series is a style of financial chart used to describe price
```
Comment (kamil-musialowski) [inline ts/Series/Renko/RenkoSeriesOptions.d.ts:29]:
```suggestion
 * A Renko series is a style of financial chart used to describe price
```

**Review by kamil-musialowski — CHANGES_REQUESTED** [review]
Minor typos, other than that looks really well 👏

**Comment by kamil-musialowski** [comment]
Switching the series type via Stock Tools also doesn't work for this series, so I'd suggest adding a new issue like this one: https://github.com/highcharts/highcharts/issues/22021

**Review by TorsteinHonsi — COMMENTED** [review]
Hey, thanks for the PR! I'm commenting early here - like the P&F series we need to refactor this to work with the v12 DataTable architecture. See the other PR for details.

File: samples/stock/demo/renko-vs-heikinashi-vs-candlestick/demo.js [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+
+    Highcharts.stockChart('container', {
+        title: {
+            text: 'Renko chart',
+            align: 'left'
+        },
+        series: [{
+            type: 'renko',
+            name: 'Renko',
+            data: data
```
Comment (kamil-musialowski) [inline samples/stock/demo/renko-vs-heikinashi-vs-candlestick/demo.js:16]:
```suggestion
            data
```

File: samples/stock/demo/renko-vs-heikinashi-vs-candlestick/demo.js [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+                text: 'Heikin Ashi'
+            },
+            top: '50%',
+            height: '50%',
+            offset: 0
+        }],
+        series: [{
+            type: 'candlestick',
+            name: 'Candlestick',
+            data: data
```
Comment (kamil-musialowski) [inline samples/stock/demo/renko-vs-heikinashi-vs-candlestick/demo.js:44]:
```suggestion
            data
```

File: samples/stock/demo/renko-vs-heikinashi-vs-candlestick/demo.js [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+        }, {
+            type: 'heikinashi',
+            name: 'Heikin Ashi',
+            data: data,
```
Comment (kamil-musialowski) [inline samples/stock/demo/renko-vs-heikinashi-vs-candlestick/demo.js:48]:
```suggestion
            data
```

File: samples/stock/plotoptions/renko-boxsize/demo.js [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+
+    Highcharts.stockChart('container', {
+        rangeSelector: {
+            selected: 1
+        },
+        series: [
+            {
+                boxSize: 6,
+                type: 'renko',
+                yAxis: 0,
```
Comment (kamil-musialowski) [inline samples/stock/plotoptions/renko-boxsize/demo.js:14]: Do we need this `yAxis`? 🤔

File: ts/Series/Renko/RenkoSeries.ts [hunk 23]
```diff
@@ -0,0 +1,228 @@
...
+    public getProcessedData(): Series.ProcessedDataObject {
+        const modified = this.dataTable.modified;
+        const processedXData: number[] = [];
+        const processedYData: number[] = [];
+        const processedLowData: number[] = [];
+        const xData = this.getColumn('x', true);
+        const yData = this.getColumn('y', true);
+        if (!this.renkoData || this.renkoData.length > 0) {
+            return {
+                modified: this.dataTable.modified,
```
Comment (kamil-musialowski) [inline ts/Series/Renko/RenkoSeries.ts:96]:
```suggestion
                modified,
```

File: ts/Series/Renko/RenkoSeries.ts [hunk 24]
```diff
@@ -0,0 +1,228 @@
...
+        }
+
+        this.processedData = renkoData;
+
+        modified.setColumn('x', processedXData);
+        modified.setColumn('y', processedYData);
+        modified.setColumn('low', processedLowData);
+
+        return {
+            modified: modified,
```
Comment (kamil-musialowski) [inline ts/Series/Renko/RenkoSeries.ts:164]:
```suggestion
            modified,
```

File: ts/Series/Renko/RenkoSeriesDefaults.ts [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
```
Comment (kamil-musialowski) [inline ts/Series/Renko/RenkoSeriesDefaults.ts:29]:
```suggestion
 * A Renko series is a style of financial chart used to describe price
```

**Review by kamil-musialowski — CHANGES_REQUESTED** [review]

File: docs/stock/renko.md [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+    renko: {
+        boxSize: 1   // This signifies that each "brick" represents $1 price changes
+    }
+}
+```
+Time is not a factor in Renko charts but Highcharts allows you to use a `datetime` axis to indicate a timeline for your price changes, offering a temporal perspective.
+
+## Highcharts Navigator Series
+
+By default, for a brief overview of the price data, smaller Renko chart is beneath the main chart using the `navigator` feature:
```
Comment (TorsteinHonsi) [inline docs/stock/renko.md:37]:
```suggestion
By default, for a brief overview of the price data, a smaller Renko chart is beneath the main chart using the `navigator` feature:
```

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Thanks, nice work! Only one minor grammatical issue from my end.

**Review by pawelfus — CHANGES_REQUESTED** [review]
I don't see `series.data` anymore in API ref. The other options are displayed properly:
Zrzut ekranu 2024-11-4 o 12 36 17

**Review by kamil-musialowski — APPROVED** [review]

**Review by DJTechnoo — APPROVED** [review]
👏

**Review by pawelfus — APPROVED** [review]
Thanks!!

**Review by TorsteinHonsi — APPROVED** [review]

**Pull request merged**